#5908: Print module documentation (#5909)
geosolutions-it/MapStore2: docs/developer-guide/printing-module.md

# Printing Module

The **printing module** of MapStore is a back-end service **not included  by default in the release package** that allows to create a printable PDF from the current map.

!!! note

>     The **printing module** is required by the **Print plugin** of MapStore, so if you want to have the Print plugin working in you application, you have to include also the printing module in your MapStore installation.

Reviewer: ```suggestion
    The **printing module** is required by the **Print plugin** of MapStore, so if you want to have the Print plugin working in your application, you have to include also the printing module in your MapStore installation.
```

## Including the printing module in MapStore

> Because MapStore doesn't include the printing module by default, to use it you need to build rom the source a `MapStore.war` that includes it or add the missing files to an existing `MapStore.war`

Reviewer: recheck this sentence because it is missing something especially the part

> rom the source a `MapStore.war`

Author: An `f` is missing, I think :cry: 
```suggestion
Because MapStore doesn't include the printing module by default, to use it you need to build from the source a `MapStore.war` that includes it OR add the missing files to an existing `MapStore.war`
```

### Building from the Source

> If you want to include the printing module in your MapStore, building the source code you have to add the profile `printing` (profiles are the 2nd argument of the `build.sh` script, after the version):

Reviewer: ```suggestion
If you want to include the printing module in your MapStore, by building the source code, you have to add the profile `printing` (profiles are the 2nd argument of the `build.sh` script, after the version):
```

when you say "profiles are the 2nd" you mean you can add more profiles as second argument? if that is not the case i would change to "is the 2nd"

Author: Yes you can add more than one profile using `,` as separator.
We have only one other profile,  `release` (that creates the binary), so you can create a binary with the printing using `printing,release`

Reviewer: ok, can you consider to add this detail  ?

```bash
./build.sh [version_identifier] printing
```

> Also MapStore projects allow to use the `printing` profile to include this module.

Reviewer: ```suggestion
MapStore projects allow to use the `printing` profile to include this module.
```

Author: I wanted to express that MapStore (standard) and MapStore (projects) build scripts can both use the `printing` profile to include the necessary things. What do you thing about this @MV88 
```suggestion
MapStore projects also allow to use the `printing` profile to include this module. So you can use the same `printing` profile to build your custom MapStore project including the printing module.
```

### Adding to an existing MapStore

It is also possible to build only the printing extension as a zip (to be unzipped on your deployed MapStore, inside the `WEB-INF/lib` directory). running `mvn clean install -Pprintingbundle`

> The zip bundle will be in `printing/target/mapstore-printing.zip`.

Reviewer: i would describe what it means by unzip, because it is not fully clear to me, 
you mean to unzip the whole zip or just the libraries (jar)?

have you checked the versions of spring ?
which the recommended one to use?

maybe you can follow up [this](https://groups.google.com/g/mapstore-developers/c/XZx3hOLrmMU/m/IVouI9npAgAJ) email in the google group

Reviewer 2: you have to unzip the whole zip to the webapps/mapstore folder of your java container

To avoid this problem, you can use several solution, depending on your setup and your reverse proxy.

#### Setting up your proxy

Some typical solutions are:

* Using AJP instead of http (this forwards all the information by default)

```conf
# Example for Apache HTTP server
ProxyPass        /mapstore   ajp://localhost:8010/mapstore
ProxyPassReverse /mapstore   ajp://localhost:8010/mapstore
```

* Using rewrite engine to rewrite the requests. (apache web server)

```conf
# example for Apache HTTP server
RewriteEngine On
RewriteCond %{HTTPS} off
RewriteRule ^ https://%{HTTP_HOST}%{REQUEST_URI}
RedirectMatch ^/$ /mapstore/
ProxyRequests Off
ProxyPreserveHost On
ProxyVia full
```

* Using non-standard headers, `X-Forwarded-Host` and `X-Forwarded-Proto`.

```conf
# example for nginx
proxy_set_header X-Forwarded-Proto https;
```

#### Forcing PRINT_BASE_URL of printing module

> Sometimes you do not have access to the . For or MapStore printing module provides a system variable `PRINT_BASE_URL` that you can set to force the URLs returned by `info.json` to be resolved from this.

Reviewer: for the ???

Author: ```suggestion
If, for any reason, you can not modify the proxy configuration, MapStore printing module provides a system variable `PRINT_BASE_URL` that you can set to force the URLs returned by `info.json` to be resolved from it.
```
